### ISSUE TITLE
SYS_getrandom undeclared compiling nim 1.6.0 stdlib on linux kernel < 3.17 (including RHEL7)

### ISSUE THREAD

**Issue by dragonfyre13** [nim-lang/Nim#19052]
On kernel versions < 3.17, compilation of nim fails at the point of compiling std/sysrand with a message similar to the below:
```
/opt/src/nim/testament_r/stdlib_sysrand.nim.c: In function ‘urandomInternalImpl__stdZsysrand_17’:
/opt/src/nim/testament_r/stdlib_sysrand.nim.c:61:35: error: ‘SYS_getrandom’ undeclared (first use in this function); did you mean ‘SYS_getuid’?
    readBytesX60gensym0_ = syscall(SYS_getrandom, ((void*) ((&dest[result]))), ((int) ((NI)(TM__sT0fWggTrXhnxQ79bl9a8nHg_2))), ((unsigned int) 0));
```

The reason for this becomes obvious pretty quickly if you're familiar with getrandom in the kernel, since it doesn't exist until version 3.17. PR is incoming with a _very_ simple fix, resolving the issue by avoiding use of getrandom when passed an associated compile flag (more on this below).

Detailed Description:

As of now, Linux distros that use older kernels (such as 3.10 in RHEL7 for example), cannot compile and use nim versions >= 1.6.0. This is because on linux kernel versions < 3.17, most notably including any/all RHEL (and CentOS) 7.x based systems, the syscall for getrandom doesn't exist. The latest stable release (1.6.0) includes sysrand in stdlib, and the missing the syscall to getrandom in the 3.10 kernel appears to be the _sole_ issue preventing functional nim compilation and usage by users of this OS/distro (presuming that GCC 8 is used via `yum install devtoolset-8` and associated initialization is performed via `scl enable devtoolset-8 bash`).

Fallback logic exists in std/sysrand which uses /dev/urandom as a PRNG, however this fallback logic is only performed if nim does not recognize it's being compiled on Linux (the logic is done for any unknown POSIX OS). If, as is the case here, nim does recognize it's being compiled on Linux, it simply uses the template for linux and fails on a missing enum for the syscall of getrandom. The fallback logic, however, provides a ready made path to resolve the issue. It should (IMHO, see below) be put behind a compile flag. This also has the added benefit of causing effectively no potential impact to already functional code/build systems.

Of course, getrandom is used because it's a cryptographically secure PRNG (Pseudo Random Number Generator) along with /dev/urandom, though it's preferred for several reasons over using /dev/urandom (for examples see [here](https://lwn.net/Articles/606141/) at the introduction of getrandom for some), however usage of /dev/urandom remains the fallback for a POSIX OS which can't be relied upon to include getrandom. Determining the presence of getrandom at runtime may be possible (I'll admit I started down that road, had several unexpected issues with runtime detection of var declaration via importc, then thought better of the entire approach). In this case, I'm not convinced it would be preferable to detect the presence of the getrandom syscall at runtime given it's intended for use as a cryptographically secure PRNG. I for one can't say that a runtime determination wouldn't expose an exploit of some kind, given I'd be doing so using likely a somewhat hamfisted approach, whereas a compile flag instead explicitly (on linux) either uses getrandom or uses the fallback of /dev/urandom. A compile flag also guarantees consistency in PRNG logic for a given binary, something that is again likely preferable when considering it's a PRNG used for cryptographic purposes (i.e. crashing on getrandom access vs falling back to /dev/urandom).

If we wanted to make this dynamic despite the above, I'll need a bit of guidance on how best to do so. When declaring a var via importc as with `var SYS_getrandom {.importc: "SYS_getrandom", header: "<sys/syscall.h>".}: clong`, subsequent calls to `declared(SYS_getrandom)` return true, however actually attempting to use the variable (i.e. `if SYS_getrandom == 0:`) throws an undeclared identifier error. As previously stated, I'd suggest it's actually preferable to have this behind a compile flag rather than a runtime or automatic determination at compile time, but I can reach out to discuss how to resolve this if my opinion differs from that of Araq or others.

**Comment by dragonfyre13** [comment]
I'll apologize in advance if I used the backport tag improperly. It seems based on my understanding as if it should be there, since this should likely be included in whichever is the next update to the 1.6.x branch, given it then allows compilation of nim on kernels < 3.17 for the most recent release of nim. Since I expect earlier versions compile fine (they don't include references to the getsystem syscall since they don't include std/sysrand), the backport tag seemed appropriate going back to 1.6 (so a future 1.6.x release would include it, thereby maintaining compatibility with kernels < 3.17 for the latest patch release in a given minor release of nim).

**Comment by Araq** [comment]
> I'll apologize in advance if I used the backport tag improperly. 

No, you didn't and even if you did, don't worry. Backporting is a manual process and we would catch mistakes.

**Comment by Stuffe** [comment]
This does not seem to be fixed in 1.6.2 as the blog post advertised.
I still get this error on Ubuntu 14.02, with the exact same error message as posted above.

**Comment by ghost** [comment]
@stuffe

> This does not seem to be fixed in 1.6.2 as the blog post advertised. I still get this error on Ubuntu 14.02, with the exact same error message as posted above.

The fix didn't change the default behaviour but added a new compile time flag - `-d:nimNoGetRandom`

**Comment by Stuffe** [comment]
> @Stuffe
> 
> > This does not seem to be fixed in 1.6.2 as the blog post advertised. I still get this error on Ubuntu 14.02, with the exact same error message as posted above.
> 
> The fix didn't change the default behaviour but added a new compile time flag - `-d:nimNoGetRandom`

Ok I see, thank you
